### PR TITLE
Deprecate assigning dict object to signals

### PR DESCRIPTION
**Pull request by ktbarrett**
This custom format `dict` value is not documented anywhere, and it's an operation that can be easily done by the user. The format was of the form `dict(values=[0, 1, 2], bits=4)`, and represents `values` as digits of bit size `bits`.

### CODE REVIEW

**Comment by eric-wieser** [comment]
I don't find the replacement particularly compelling - I'd be inclined to leave this behavior in, and try to design a replacement as part of the new `BinaryValue`. In this case, it seems what's needed is a way to view an `m*n`-bit signal as an array of `m`-bit signals.

**Comment by ktbarrett** [comment]
@eric-wieser If we plan on replacing it with something else in the next version, this interface will break, so this interface should be deprecated.

**Comment by ktbarrett** [comment]
As for a proper replacement, this could be a feature of the `Vector` class, where vectors of vectors of Logic values can be flattened. This behavior doesn't really belong in the assignment operator, which should be kept as simple as possible.

**Comment by eric-wieser** [comment]
Perhaps a compromise is to introduce in `cocotb.utils`:
```python
# or a better name
def pack_bit_vector(values : List[int], *, bits : int):
    """ Pack the integers in `values` into a single integer, with each entry occupying `bits` bits. """
    return sum(v << (bits * i) for i, v in enumerate(values))
```
which means:

* We can deprecate the weird dict thing now, and recommend something short but less surprising
* We can deprecate `pack_bit_vector` in 2.0 when we gut binaryvalue.

File: cocotb/handle.py [hunk 1]
```diff
@@ -698,6 +698,11 @@ def _set_value(self, value, call_sim):
         Raises:
             TypeError: If target is not wide enough or has an unsupported type
                  for value assignment.
+
+        .. deprecated:: 1.5
+            :class:`dict` objects are no longer accepted as values for assignment.
+            Convert the dictionary to an integer before assignment using
+            ``sum(v << (d['bits'] * i) for i, v in enumerate(d['values']))``.
```
Comment (eric-wieser) [inline cocotb/handle.py:705]: This implementation isn't the same as the one we are deprecating, it reverses the entries!

Comment (ktbarrett) [reply cocotb/handle.py:705]: That's due to how it's building the integer. I got tricked by using `reversed` at first. The test shows it's correct. But I will do what you had mentioned above.

**Review by eric-wieser — COMMENTED** [review]
The suggested replacement is less safe: `if len(vallist) * value["bits"] != len(self):` is present in the original codepath.

**Comment by ktbarrett** [comment]
@eric-wieser So is the suggestion to document an assert in the recommended replacement, or to flip this change on it's head and document the interface since it's potentially useful? I don't particularly care. It's undocumented, so I think it can be removed with prejudice. Even if we keep it, it will live only to be deprecated once we have proper logic vector  value assignment support.

**Pull request closed without merging**